File: frontend/services/warranty.ts
import apiClient from "@/services/api";
import { GlobalResponse } from "@/services/type/responses";

export type WarrantyStatus = "inactive" | "active" | "expired"

export type Warranty = {
    ID: number;
    product_sn: string;
    product_type: string;
    product_model: string;
    activation_date: string;
    expiry_date: string;
    do_date: string;
    warranty_length: number;
    warranty_status: string;
    name: string;
    email: string;
    phone_number: string;
    province: string;
    city: string;
}

export type CheckWarrantyResult = {
    product_sn: string;
    product_type: string;
    product_model: string;
    activation_date: string;
    expiry_date: string;
    warranty_length: string;
    status: string;
    remaining_days: number;
    name: string;
}

export type ActivateWarrantyResult = {
    expiry_date: string;
    remaining_days: string;
    product_sn: string;
    status: string;
}

export type ActivateWarrantyDTO = {
    name: string
    email: string
    phone_number: string
    province: string
    city: string
    product_sn: string
}

export type WarrantyParams = {
    limit: number;
    page: number;
    q: string;
}

export const getAllWarranties = async (searchParams: WarrantyParams) => {
    return apiClient.get<GlobalResponse<Warranty[]>>(`/api/warranties/`, { params: searchParams })
}

export const CheckProductWarranty = async (productSN: string) => {
    return apiClient.get<GlobalResponse<CheckWarrantyResult>>(
        `/api/warranties/${productSN}`
    )
}

export const ActivateWarranty = async (payload: ActivateWarrantyDTO) => {
    return apiClient.post<GlobalResponse<ActivateWarrantyResult>>(
        `/api/warranties/${payload.product_sn}`, payload
    )
}

export const importWarranty = async (file: File) => {
    const formData = new FormData();
    formData.append("file", file);

    return apiClient.post<GlobalResponse<any>>(`/api/warranties/import`, formData, {
        headers: {
            "Content-Type": "multipart/form-data",
        },
    });
};

File: backend/models/warranty.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type WarrantyStatus string

const (
	WarrantyActive   WarrantyStatus = "active"
	WarrantyExpired  WarrantyStatus = "expired"
	WarrantyInactive WarrantyStatus = "inactive"
)

type Warranty struct {
	gorm.Model
	ProductSN    string `gorm:"uniqueIndex" json:"product_sn"`
	ProductType  string `gorm:"not null" json:"product_type"`
	ProductModel string `gorm:"not null" json:"product_model"`

	ActivationDate *time.Time     `json:"activation_date"`
	ExpiryDate     *time.Time     `json:"expiry_date"`
	DODate         *time.Time     `json:"do_date"`
	WarrantyLength int            `gorm:"not null" json:"warranty_length"` // in days
	WarrantyStatus WarrantyStatus `gorm:"type:varchar(20);check:warranty_status IN ('active','expired','inactive');default:'inactive'" json:"warranty_status"`

	Name        string `json:"name"`
	Email       string `json:"email"`
	PhoneNumber string `json:"phone_number"`
	Province    string `json:"province"`
	City        string `json:"city"`
}
